### ISSUE TITLE
asm (keystone) command doesn't provide valid results after switching to ARM THUMB mode

### ISSUE THREAD

**Issue by mkopec3** [hugsy/gef#303]
Your issue will be closed unless you confirm the following:
  * [X] Did you use the latest version of GEF from `master` branch?
  * [X] Did you read the [documentation](https://gef.readthedocs.org/en/latest/) first?
  * [X] Did you check [closed issues](https://github.com/hugsy/gef/issues)?

### Step 1: Describe your environment

  * Operating System: Linux arm-v6l-pi 4.1.19 #1 PREEMPT Tue Mar 15 18:12:17 CDT 2016 armv6l GNU/Linux
  * Architecture: armv6l
  * GDB version (including the Python library version): GDB 8.1 using Python engine 3.6

#### Steps to reproduce

The following code can be used to reproduce the issue. Check the comments for the steps required for reproduction.

```
// thumb_demo.s
// Have to be built and ran on an ARM platform
// as thumb_demo.s -o thumb_demo.o
// ld thumb_demo.o -o thumb_demo
// gdb -q ./thumb_demo
// b _start
// r
// n 2
// asm "mov r0, #1"

.section .text
.global _start

_start:
        .code 32
        add r3, pc, #1	// run asm "mov r0, #1", that provides valid results
        bx  r3		// entering into ARM THUMB mode

        .code 16
        mov r0, #1	// run asm "mov r0, #1", that DOESNT' PROVIDE VALID results (not Thumb code)
        mov r1, #2
        add r2, r0, r1
```

#### Observed Results

  * asm (keystone-assembler) command seems to not recognize that ARM CPU switched into THUMB mode, therefore producing wrong instruction set

#### Expected results

  * GEF should identify that ARM switched into THUMB mode (based on the value in CPSR register), and then appropriately call keystone interface (providing valid arch+mode value)

#### Traces

![gef_keystone_asm_bug](https://user-images.githubusercontent.com/25580985/44255416-55ee8a80-a206-11e8-83fa-6e3fb220cc7e.png)

Thumb and ARM code produced from the same mnemonic set by kstool:

![kstool_thumb](https://user-images.githubusercontent.com/25580985/44255555-c39ab680-a206-11e8-8671-2efef946c6a3.png)

The issue described here is similar to issue #302.
Added as a separate issue for clarity.

**Comment by hugsy** [comment]
Noted (regarding the current running architecture), but FYI:

```
gef➤  asm -a arm -m arm mov r0, #1
[+] Assembling 1 instruction for ARM:ARM (little endian)
\x01\x00\xa0\xe3                                             # mov r0, #1
gef➤  asm -a arm -m thumb mov r0, #1
[+] Assembling 1 instruction for ARM:THUMB (little endian)
\x4f\xf0\x01\x00                                             # mov r0, #1
```

**Comment by hugsy** [comment]
Fixed by 4b0e110d5286fd8385df00f5781a85335d4ecced (in `dev` until 2018.10 is released - soon).
